net/rhp/merkle.go

```diff
@@ -251,6 +256,107 @@ func nextSubtreeSize(start, end int) int {
+// SegmentsPerSector/2. If a precalculated root is not available, precalc
+// should return the zero hash.
+func BuildProof(sector *[SectorSize]byte, start, end int, precalc func(i, j int) types.Hash256) []types.Hash256 {
+	if start < 0 || end > LeavesPerSector || start > end || start == end {
+		panic("BuildProof: illegal proof range")
+	}
+	if precalc == nil {
+		precalc = func(i, j int) (h types.Hash256) { return }
+	}
+
+	// define a helper function for later
+	var s sectorAccumulator
+	subtreeRoot := func(i, j int) types.Hash256 {
+		s.reset()
+		s.appendLeaves(sector[i*LeafSize : j*LeafSize])
+		return s.root()
+	}
+
+	// we build the proof by recursively enumerating subtrees, left to right.
+	// If a subtree is inside the segment range, we can skip it (because the
+	// verifier has the segments); otherwise, we add its Merkle root to the
+	// proof.
+	//
+	// NOTE: this operation might be a little tricky to understand because
+	// it's a recursive function with side effects (appending to proof), but
+	// this is the simplest way I was able to implement it. Namely, it has the
+	// important advantage of being symmetrical to the Verify operation.
+	proof := make([]types.Hash256, 0, ProofSize(LeavesPerSector, start))
+	var rec func(int, int)
+	rec = func(i, j int) {
+		if i >= start && j <= end {
+			// this subtree contains only data segments; skip it
+		} else if j <= start || i >= end {
+			// this subtree does not contain any data segments; add its Merkle
+			// root to the proof. If we have a precalculated root, use that;
+			// otherwise, calculate it from scratch.
+			if h := precalc(i, j); h != (types.Hash256{}) {
+				proof = append(proof, h)
+			} else {
+				proof = append(proof, subtreeRoot(i, j))
+			}
+		} else {
+			// this subtree partially overlaps the data segments; split it
+			// into two subtrees and recurse on each
+			mid := (i + j) / 2
+			rec(i, mid)
+			rec(mid, j)
+		}
+	}
+	rec(0, LeavesPerSector)
+	return proof
+}
+
+// BuildSectorRangeProof constructs a proof for the sector range [start, end).
+func BuildSectorRangeProof(sectorRoots []types.Hash256, start, end int) []types.Hash256 {
+	if len(sectorRoots) == 0 {
+		return nil
+	} else if start < 0 || end > len(sectorRoots) || start > end || start == end {
+		panic("BuildSectorRangeProof: illegal proof range")
+	}
+
+	proof := make([]types.Hash256, 0, ProofSize(len(sectorRoots), start))
+	buildRange := func(i, j int) {
+		for i < j && i < len(sectorRoots) {
+			subtreeSize := nextSubtreeSize(i, j)
+			if i+subtreeSize > len(sectorRoots) {
+				subtreeSize = len(sectorRoots) - i
+			}
+			proof = append(proof, MetaRoot(sectorRoots[i:][:subtreeSize]))
+			i += subtreeSize
+		}
+	}
+	buildRange(0, start)
+	buildRange(end, math.MaxInt32)
+	return proof
+}
+
+// ConvertProofOrdering converts "left-to-right" proofs into the "leaf-to-root"
+// ordering used in consensus storage proofs.
+func ConvertProofOrdering(proof []types.Hash256, index int) []types.Hash256 {
```

Comment:
I don't think we need this anymore. Sia v1 validates storage proofs in "leaf-to-root" order, whereas Sia v2 validates them "left-to-right" (matching how all other proofs are constructed in the RHP).